### DISCUSSION TITLE
Can you open a folder in a Codespace?

### DISCUSSION THREAD

**Discussion by bmalehorn** [community/community#5181] [Codespaces]
Normally, I can open any folder in VSCode, then re-open it in a development container:

```
git clone git@github.com:bmalehorn/test-codespaces-public.git
code test-codespaces-public/service1
```
<img width="1136" alt="Screen Shot 2021-08-18 at 4 04 33 PM" src="https://user-images.githubusercontent.com/3344958/129983342-a159f786-dd59-471a-8796-985039bdf41d.png">

This runs the VSCode command `Remote-Containers: Open Folder in Container...`.

But with Codespaces, there's no option to pick what folder to open:

<img width="1218" alt="Untitled" src="https://user-images.githubusercontent.com/3344958/129983774-538191f7-3d61-4f5c-be44-32dd69d42ef2.png">

Is it possible to open a Codespace to a particular folder in a repository? Or is it only possible to open in the root folder?

Ideally there's be a button to open a folder in a Codespace (maybe only displayed if they had a .devcontainer directory):

<img width="1129" alt="Untitled 3" src="https://user-images.githubusercontent.com/3344958/129986331-5e801767-42aa-4372-9ebf-2491c4624c76.png">

I'm asking because at my organization, we have 1 repository containing 100+ services. It's not realistic make 1 devcontainer image that will come pre-installed with all dependencies of all services, or to create 1 VSCode configuration that works across all services.

**Comment by tanmayeekamath** [comment]
Hey @bmalehorn! Thanks for your feedback. Currently we support selecting the folder through a devcontainer config as you pointed out correctly. 
However, @2percentsilk is looking into how we can support Codespaces for a monorepos where there might be need to have more than one devcontainer for each service inside a single repo and this feedback is extremely helpful in thinking through different scenarios/needs.

**Reply by bmalehorn** [reply]
Thanks for the confirmation. It seems like the technology is all there, but it will require some UI & bookkeeping to work to make it a good user experience.

**Comment by radu-matei** [comment]
While it's not possible to open a single directory, you could add the intended directory to the VS Code workspace.

**Comment by DonKirkham** [comment]
I love Codespaces and this is my #1 request.  I work with the repos in the @pnp organization, which consists of hundreds on sample projects for developers. We are currently adding the appropriate .devcontainer files to each project because the project range over a dozen versions of SharePoint Framework, many of which require a specific versions of devtools, like nodeJs, npm, gulp, SPFx, etc.  If I clone this to my local machine and open the workbench in a project folder, I have the option to use remote-containers (Docker) to load my project in a pre-configured environment.  It would be awesome to have this ability directly in Codespaces.

Currently, i open the monorepo in Codespaces, then reopen VS Code in a project folder.  I then use nvm to install the needed version of nodeJs and the rest of the tools I need. As long as I keep the codespace, I can keep doing this, and reuse environments that match my requirements. If I delete the codespace (saving $$$), I have to start over.

Since the full repo is cloned into the VM before the container is created, if there was an option to specify the starting folder when the Codespace is created (or let a Codespace be created from inside a folder in Github), the .devcontainer files could be used to build the container instead of the root .devcontainer. This would be the simplier approach, but if the VS Code extension could be updated to recognize that when you reopen VS COde in a folder, you are already running in Codespaces and offer to recreate (add?) a container based on the local .devcontainer files.

Happy to discuss this more with the Product Group or PMs. Thanks and keep up the great work.

**Comment by bartr** [comment]
From the CS terminal, run `code /workspaces/foo`

A new tab will open up with /workspaces/foo as the root (sometimes you have to run it twice)

`code $HOME` or `code ~` works too

`code *.md` also works

If you get a popup in your original window, just click cancel.

On our current project, we clone 8 different repos to /workspaces and change between them often. It works really well (almost all the time - an occasional "wierd" behavior)

**Comment by craiglpeters** [comment]
Configuration support for monorepos is now supported - see https://docs.github.com/en/enterprise-cloud@latest/codespaces/setting-up-your-project-for-codespaces/introduction-to-dev-containers#devcontainerjson
